### ISSUE TITLE
ManyToMany relations in session [Jelly Collection doesn't serialize properly]

### ISSUE THREAD

**Issue by SpadXIII** [jonathangeiger/kohana-jelly#71]
I've just been struggling with Jelly-auth and I've ran into a little problem. (I've also just messaged Raeldc through here about this)

When I'm using the user_roles (which is a ManyToMany relation) it works fine at first, then after a refresh, the `$user->roles` is `NULL` instead of the actual roles. I think it has something to do with the `$user` object being stored (serialized?) in the Session and then later on retrieved.

I haven't tested this thoroughly enough yet, I've just been banging my head against the wall because it doesn't work..

**Comment by banks** [comment]
> I think it has something to do with the $user object being stored (serialized?) in the Session and then later on retrieved.

Almost certainly although in theory that should work fine due to our implementation of `__sleep()` in `Jelly_Model`. Can you pastie the debug output of your whole user model after restoring it from the session?

**Comment by SpadXIII** [comment]
`__sleep` in `Jelly_Model`? I didn't see any, though I did see a `__sleep` in `Jelly_Collection`

Here's a paste of the user object right after login (good) and after a refresh (not good) in a debug echo:

[pastie](http://pastie.org/private/ww1zkiafwdds4lll3pgmta)

The roles are NULL after the refresh. No idea how/why/whatever .. I haven't tested this with other objects containing ManyToMany-relations (yet?).

**Comment by banks** [comment]
> __sleep in Jelly_Model? I didn't see any, though I did see a __sleep in Jelly_Collection

My mistake. It used to be in the model. In fact i wonder if this could be a contributing factor to why this doesn't work.

Can you check whether it is purely a serialization issue? Right after login do this:

```
Kohana:debug(unserialize(serialize($user)));
```

And see if the same thing happens. I'm not sure why the $_retrieved array is different in the object from the session.

**Comment by SpadXIII** [comment]
I think there's more going on here. Maybe it's a better idea if you do some debugging yourself, as you know what you're looking for.

At first, the `unserialize(serialize($user))` worked. But that was before accessing any of the roles. So I accessed the `$user->roles` right above the un-/serialize line. And that gave an error.. a very clear error: `Fatal error: Exception thrown without a stack frame in Unknown on line 0` ..

I've added an `exit;` after the echo-line and that gives a different error:

```
ErrorException [ Notice ]: serialize() [function.serialize]: __sleep should return an array only containing the names of instance-variables to serialize
```

The code used in my test:

```
if ($this->user->roles)
{
    echo 'has roles';
}
echo Kohana::debug('USER', $this->user);
echo '<br/><br/>';
echo Kohana::debug('UNSERIALIZED', unserialize(serialize($this->user)));
exit;
```

This results in this: http://pastie.org/private/c7ldqzdedjll9nz6hlxpyw

**Comment by banks** [comment]
> Fatal error: Exception thrown without a stack frame in Unknown on line 0

This happens when an exception is thrown in a method call ed in __destruct. In this case, the error is while serilaising the object in `Session::__destruct()`. That makes sense - basically the `Jelly_Collection::__sleep()` is returning something invalid (nothing) and since the error isn't handled properly (because it happens after script execution has finished and things are being dismantled) I guess you just get NULL.

The problem is that `Jelly_Collection::__sleep()` is returning nothing I guess - this clearly hasn't been tested since this code was moved form `Jelly_Model`.

We will fix it. I guess the correct line would be something like:

```
return array_keys(get_object_vars($this));
```

I may not get a chance to test this properly for a few days though so if you can confirm it works that would be a help.

**Comment by banks** [comment]
Made Jelly Collections serializable. Closed by cd727fe75d3b4888fed502b7d508a56a2b2bfceb
